autogen/oai/client.py

```diff
@@ -52,6 +54,209 @@
 LEGACY_CACHE_DIR = ".cache"
 
 
+class Client(ABC):
+    """
+    A client class must implement the following methods:
+    - create must return a response object that implements the ClientResponseProtocol
+    - cost
+
+    This class is used to create a client that can be used by OpenAIWrapper.
+    It mimics the OpenAI class, but allows for custom clients to be used.
+    """
+
+    RESPONSE_USAGE_KEYS = ["prompt_tokens", "completion_tokens", "total_tokens", "cost", "model"]
+
+    class ClientResponseProtocol(Protocol):
+        class Choice(Protocol):
+            class Message(Protocol):
+                content: str | None
+                function_call: str | None
+
+        choices: List[Choice]
+        config_id: int
+        cost: float
+        pass_filter: bool
+        model: str
+
+    def update(self, config: Dict):
+        # update with anything here
+        pass
+
+    @abstractmethod
+    def create(self, params) -> ClientResponseProtocol:
+        pass
+
+    @abstractmethod
+    def cost(self, response: ClientResponseProtocol) -> float:
+        pass
+
+    @staticmethod
+    def get_usage(response: ClientResponseProtocol) -> Dict:
+        return None
+
+
+class OpenAIClient(Client):
```

Comment:
If `Client` is a `Protocol`, then there is no need to inherit from it.

Reply:
updated